### PR TITLE
TriGParser: distinguish empty entities from no-entity being read

### LINKED ISSUE #37: Turtle parser fails for unknown reason (SPARQL 1.1 Syntax Update 1 manifest.ttl)
@pietercolpaert When parsing turtle file https://www.w3.org/2009/sparql/docs/tests/data-sparql11/syntax-update-1/manifest parser returns 0 triples. 

**ARC2**'s Turtle parser has no problems and returns all triples, so file should be correct.

I made a test which shows the problem: https://github.com/pietercolpaert/hardf/blob/error/turtle-parser-fails-unknown/test/TriGParserTest.php#L2061-L2072 (branch **error/turtle-parser-fails-unknown**)

Any idea why?

### PR DESCRIPTION
**Pull request by zozlak**
Solves the #37 by throwing an informative exception describing why the parsing can not be done without the document base IRI being set.

### CODE REVIEW

**Comment by zozlak** [comment]
See https://github.com/pietercolpaert/hardf/issues/37#issuecomment-1967174143

What is your (@k00ni, @pietercolpaert ) preferred way to deal with that?

I can either adjust the parser to consistently return empty subject/predicate/object on "diamonds" (`<>`) without the base IRI set (it is only a matter of turning comparisons like [this one](https://github.com/pietercolpaert/hardf/pull/42/files#diff-e6bc095837c1bd1a4cc8fb0d28982526423dadb30a702c0967185504e4f27f60L356) into type-strict ones) or I can adjust the failing test to expect the "please set the documentIRI" exception.

**Comment by pietercolpaert** [comment]
Adjusting the test to expect an exception takes my preference

**Review by pietercolpaert — APPROVED** [review]

File: test/TriGParserTest.php [hunk 1]
```diff
@@ -69,6 +69,9 @@ public function shouldNotParse($createParser, $input, $expectedError = null): vo
                 $this->fail("Expected this error to be thrown (but it wasn't): ".$expectedError);
             }
         });
+        if (!$errorReceived) {
```
Comment (k00ni) [inline test/TriGParserTest.php]:
```suggestion
        if (false === $errorReceived) {
```
Just a small thing, but its more obvious what we check.

File: test/TriGParserTest.php [hunk 2]
```diff
@@ -270,13 +273,18 @@ public function testZeroOrMoreTriples(): void
...
-        $this->shouldParse("<> <> <> <>.\n(<>) <> (<>) <>.",
-        ['', '', '', ''],
-        ['_:b0', '', '_:b1', ''],
-        ['_:b0', 'http://www.w3.org/1999/02/22-rdf-syntax-ns#first', ''],
+        // should throw an error on empty list item with lacking document base IRI
+        $this->shouldNotParse("(<>) <> (<>) <>.",
+        "list item on line 1 can not be parsed without knowing the the document base IRI.\n".
+        "Please set the document base IRI using the documentIRI parser configuration option.\n".
+        "See https://github.com/pietercolpaert/hardf/#empty-document-base-IRI .");
+        // but should manage if the parser has documentIRI set
```
Comment (k00ni) [inline test/TriGParserTest.php]:
```suggestion
        "See https://github.com/pietercolpaert/hardf/#empty-document-base-IRI .");

        // but should manage if the parser has documentIRI set
```

**Review by k00ni — CHANGES_REQUESTED** [review]
:+1: Thank you @zozlak, looks good to me, just a few remarks.

I also support @pietercolpaert's preference.

_Off topic: The current coding style should be addressed in the future, because it makes code (especially the tests) very hard to read. Therefore only a few remarks in this regard._

File: test/TriGParserTest.php (hunk 1 above)
Comment (pietercolpaert) [reply test/TriGParserTest.php]: Ufff, I disagree with testing whether a boolean is a certain boolean

**Pull request merged**